templates/commands/describe/describe.go

```diff
@@ -0,0 +1,174 @@
+
+  - "helloworld@v1" means "github.com/abcxyz/helloworld repo at revision v1;
+    this is for a template owned by abcxyz.
+  - "myorg/helloworld@v1" means github.com/myorg/helloworld repo at revision
+    v1; this is for a template not owned by abcxyz but still on GitHub.
+  - "mygithost.com/mygitrepo/helloworld@v1" is for a template in a remote git
+    repo but not owned by abcxyz and not on GitHub.
+  - "mylocaltarball.tgz" is for a template not in git but present on the local
+    filesystem.
+  - "http://example.com/myremotetarball.tgz" os for a non-Git template in a
+    remote tarball.`
+}
+
+func (c *Command) Flags() *cli.FlagSet {
+	set := c.NewFlagSet()
+	c.flags.Register(set)
+	return set
+}
+
+type runParams struct {
+	fs     common.FS
+	stdout io.Writer
+
+	describeTempDirBase string
+}
+
+func (c *Command) Run(ctx context.Context, args []string) error {
+	if err := c.Flags().Parse(args); err != nil {
+		return fmt.Errorf("failed to parse flags: %w", err)
+	}
+	fSys := c.testFS
+	if fSys == nil {
+		fSys = &common.RealFS{}
+	}
+	return c.realRun(ctx, &runParams{
+		fs:     fSys,
+		stdout: c.Stdout(),
+	})
+}
+
+// readRun provides fakable interface to test Run.
+func (c *Command) realRun(ctx context.Context, rp *runParams) (rErr error) {
+	var tempDirs []string
+	defer func() {
+		var rErr error
+		for _, d := range tempDirs {
+			rErr = errors.Join(rErr, rp.fs.RemoveAll(d))
+		}
+	}()
+	_, templateDir, err := templatesource.Download(ctx, rp.fs, rp.describeTempDirBase, c.flags.Source, c.flags.GitProtocol)
+	if templateDir != "" { // templateDir might be set even if there's an error
+		tempDirs = append(tempDirs, templateDir)
+	}
+	if err != nil {
+		return err //nolint:wrapcheck
+	}
+
+	specPath := filepath.Join(templateDir, specName)
+	f, err := rp.fs.Open(specPath)
+	if err != nil {
+		return fmt.Errorf("error opening template spec: ReadFile(): %w", err)
+	}
+	defer f.Close()
+
+	specI, err := decode.DecodeValidateUpgrade(ctx, f, specName, decode.KindTemplate)
+	if err != nil {
+		return fmt.Errorf("error reading template spec file: %w", err)
+	}
+
+	spec, ok := specI.(*spec.Spec)
+	if !ok {
+		return fmt.Errorf("internal error: spec file did not decode to spec.Spec")
+	}
+	tw := tabwriter.NewWriter(rp.stdout, 8, 0, 2, ' ', 0)
+	fmt.Fprintf(tw, "\nTemplate:\t")
+	fmt.Fprintf(tw, "\nDescription:\t%s", spec.Desc.Val)
+	fmt.Fprintf(tw, "\n")
+	for _, i := range spec.Inputs {
+		fmt.Fprintf(tw, "\nInput name:\t%s", i.Name.Val)
+		fmt.Fprintf(tw, "\nDescription:\t%s", i.Desc.Val)
+		for idx, rule := range i.Rules {
+			printRuleIndex := len(i.Rules) > 1
+			writeRule(tw, rule, printRuleIndex, idx)
+		}
+
+		if i.Default != nil {
+			defaultStr := i.Default.Val
+			if defaultStr == "" {
+				// When empty string is the default, print it differently so
+				// the user can actually see what's happening.
+				defaultStr = `""`
+			}
+			fmt.Fprintf(tw, "\nDefault:\t%s", defaultStr)
+		}
+		fmt.Fprintf(tw, "\n")
+		tw.Flush()
```

Comment:
This logic seems like it could be extracted into a helper for more independent testing.